pkg/action/action_test.go

```diff
@@ -273,6 +274,19 @@ func namedReleaseStub(name string, status release.Status) *release.Release {
 	}
 }
 
+func Test_chartNeedsAPIVersions(t *testing.T) {
```

Comment:
I can think of several false positives and false negatives, like:
```
{{ (.Capabilities).APIVersions }}
```
```
# Warning: .Capabilities.APIVersions should not be used under any circumstances
```

Reply:
@astellingwerf Ah, thanks for pointing out these edge cases.

The function definitely needs to avoid false negatives, but I think some false positives are tolerable.

Maybe it should be updated to:
* Check for just `.APIVersions` instead of `.Capabilities.APIVersions` to avoid false negatives
* Ignore `.APIVersions` if it occurs in a comment (a line beginning with `#`)

That should be possible with a regex.

What do you think?

Reply:
The first bullet makes sense, @eolatham, but as you can see in line 2 of my second example, lines starting with `#` don't necessarily imply no actual use of capabilities.

I must admit that I've hardly seen any of the Helm code, but I'm afraid that, unless you have access to the AST of the Chart, it'll be quite a challenge to get a the test just right. I agree with your analysis that false negatives should be prevented.

Reply:
Yeah, I'd want to keep the implementation and testing as simple as possible since this is meant to be a straightforward optimization.

But should references to `.APIVersions` in comments be considered "usage" from this perspective?
I wouldn't think so because comments have no effect on the rendered output, right?

Reply:
Depends on the kind of comment. A YAML comment (though quite possibly not impacting the functioning of the rendered app) does impact the output:
```
# {{ .Capabilities.APIVersions | toJson }}
```
whereas this, with the "Helm code" commented out, doesn't:
```
{{/*{{ .Capabilities.APIVersions | toJson }}*/}}
```

Reply:
Right.

So I guess, if we want to consider "usage of `.APIVersions`" as any reference that alters the rendered output of the templates (including references that have no functional app impact, as you pointed out), we would have to include references to `.APIVersions` in comments as usage.

In that case, maybe the best solution is to simply check if templates contain `.APIVersions`, and avoid adding any other complexity to the logic. I think we can tolerate false positives from comments and other non-functional references since those are probably rare, and this change would still be a nice optimization for many charts.

If you agree, I'll go ahead and update the PR.

Reply:
I'm not an expert in any way (I just stumbled upon this PR because I was facing different issues with `helm template --api-versions` myself), but that makes sense.

Reply:
...or do we wanna support an edge case like:
```
{{ get (.Capabilities | toJson | fromJson) "APIVersions" | has "batch/v1" }}
```

Reply:
Nice catch, I think we have to support that too. I just updated the PR.